packages/realm/src/List.ts

```diff
@@ -36,27 +43,31 @@ type PartiallyWriteableArray<T> = Pick<Array<T>, "pop" | "push" | "shift" | "uns
  * only store values of a single type (indicated by the `type` and `optional`
  * properties of the List), and can only be modified inside a {@link Realm.write | write} transaction.
  */
-export class List<T = unknown> extends OrderedCollection<T> implements PartiallyWriteableArray<T> {
+export class List<T = unknown>
+  extends OrderedCollection<T, [number, T], ListAccessor<T>>
+  implements PartiallyWriteableArray<T>
+{
   /**
    * The representation in the binding.
    * @internal
    */
-  public declare internal: binding.List;
+  public declare readonly internal: binding.List;
 
   /** @internal */
   private declare isEmbedded: boolean;
 
   /** @internal */
-  constructor(realm: Realm, internal: binding.List, helpers: OrderedCollectionHelpers) {
+  constructor(realm: Realm, internal: binding.List, accessor: ListAccessor<T>) {
     if (arguments.length === 0 || !(internal instanceof binding.List)) {
       throw new IllegalConstructorError("List");
     }
-    super(realm, internal.asResults(), helpers);
+    const results = internal.asResults();
+    super(realm, results, accessor);
 
     // Getting the `objectSchema` off the internal will throw if base type isn't object
-    const baseType = this.results.type & ~binding.PropertyType.Flags;
     const isEmbedded =
-      baseType === binding.PropertyType.Object && internal.objectSchema.tableType === binding.TableType.Embedded;
+      toItemType(results.type) === binding.PropertyType.Object &&
+      internal.objectSchema.tableType === binding.TableType.Embedded;
```

Comment:
swapping the order might improve performance a bit 😄

Reply:
There's a comment above about `internal.objectSchema.tableType` throwing if we don't check for the Object property type first.